brilift/src/translator.rs

```diff
@@ -0,0 +1,1073 @@
+use crate::rt;
+use bril_rs as bril;
+use core::mem;
+use cranelift_codegen::entity::EntityRef;
+use cranelift_codegen::ir::condcodes::{FloatCC, IntCC};
+use cranelift_codegen::ir::InstBuilder;
+use cranelift_codegen::settings::Configurable;
+use cranelift_codegen::{ir, isa, settings};
+use cranelift_frontend::{FunctionBuilder, FunctionBuilderContext, Variable};
+use cranelift_jit::{JITBuilder, JITModule};
+use cranelift_module::{default_libcall_names, Module};
+use cranelift_object::{ObjectBuilder, ObjectModule};
+use enum_map::{enum_map, Enum, EnumMap};
+use std::collections::HashMap;
+use std::fs;
+use std::sync::Arc;
+
+/// Runtime functions used by ordinary Bril instructions.
+#[derive(Debug, Enum)]
+#[allow(clippy::enum_variant_names)]
+enum RTFunc {
+    PrintInt,
+    PrintBool,
+    PrintFloat,
+    PrintChar,
+    PrintSep,
+    PrintEnd,
+    Alloc,
+    Free,
+}
+
+impl RTFunc {
+    fn sig(
+        &self,
+        pointer_type: ir::Type,
+        call_conv: cranelift_codegen::isa::CallConv,
+    ) -> ir::Signature {
+        match self {
+            Self::PrintInt => ir::Signature {
+                params: vec![ir::AbiParam::new(ir::types::I64)],
+                returns: vec![],
+                call_conv,
+            },
+            Self::PrintBool => ir::Signature {
+                params: vec![ir::AbiParam::new(ir::types::I8)],
+                returns: vec![],
+                call_conv,
+            },
+            Self::PrintFloat => ir::Signature {
+                params: vec![ir::AbiParam::new(ir::types::F64)],
+                returns: vec![],
+                call_conv,
+            },
+            Self::PrintChar => ir::Signature {
+                params: vec![ir::AbiParam::new(ir::types::R32)],
```

Comment:
I think `R32` is a reference type… for characters, we probably want `I8`?

Reply:
Rust characters are 4 bytes wide and roughly correspond to C++ "wide characters". I had assumed that Bril's characters were the same way based on https://capra.cs.cornell.edu/bril/lang/char.html and the `bril-rs` `Literal` enum. I think that `I32` is the right thing in this case.